Mathlib/NumberTheory/LegendreSymbol/Basic.lean

```diff
@@ -0,0 +1,324 @@
+    legendreSym p a = 1 ↔ IsSquare (a : ZMod p) := by rw [eq_one_iff]; norm_cast; exact_mod_cast ha0
+#align legendre_sym.eq_one_iff' legendreSym.eq_one_iff'
+
+/-- `legendreSym p a = -1` iff `a` is a nonsquare mod `p`. -/
+theorem eq_neg_one_iff {a : ℤ} : legendreSym p a = -1 ↔ ¬IsSquare (a : ZMod p) :=
+  quadraticChar_neg_one_iff_not_isSquare
+#align legendre_sym.eq_neg_one_iff legendreSym.eq_neg_one_iff
+
+theorem eq_neg_one_iff' {a : ℕ} : legendreSym p a = -1 ↔ ¬IsSquare (a : ZMod p) := by
+  rw [eq_neg_one_iff]; norm_cast
+#align legendre_sym.eq_neg_one_iff' legendreSym.eq_neg_one_iff'
+
+/-- The number of square roots of `a` modulo `p` is determined by the Legendre symbol. -/
+theorem card_sqrts (hp : p ≠ 2) (a : ℤ) :
+    ↑{x : ZMod p | x ^ 2 = a}.toFinset.card = legendreSym p a + 1 :=
+  quadraticChar_card_sqrts ((ringChar_zmod_n p).substr hp) a
+#align legendre_sym.card_sqrts legendreSym.card_sqrts
+
+end legendreSym
+
+end Legendre
+
+section QuadraticForm
+
+/-!
+### Applications to binary quadratic forms
+-/
+
+
+namespace legendreSym
+
+/-- The Legendre symbol `legendreSym p a = 1` if there is a solution in `ℤ/pℤ`
+of the equation `x^2 - a*y^2 = 0` with `y ≠ 0`. -/
+theorem eq_one_of_sq_sub_mul_sq_eq_zero {p : ℕ} [Fact p.Prime] {a : ℤ} (ha : (a : ZMod p) ≠ 0)
+    {x y : ZMod p} (hy : y ≠ 0) (hxy : x ^ 2 - a * y ^ 2 = 0) : legendreSym p a = 1 := by
+  apply_fun (· * y⁻¹ ^ 2) at hxy
+  simp only [MulZeroClass.zero_mul] at hxy
+  rw [(by ring : (x ^ 2 - ↑a * y ^ 2) * y⁻¹ ^ 2 = (x * y⁻¹) ^ 2 - a * (y * y⁻¹) ^ 2),
+    mul_inv_cancel hy, one_pow, mul_one, sub_eq_zero, pow_two] at hxy
+  exact (eq_one_iff p ha).mpr ⟨x * y⁻¹, hxy.symm⟩
+#align legendre_sym.eq_one_of_sq_sub_mul_sq_eq_zero legendreSym.eq_one_of_sq_sub_mul_sq_eq_zero
+
+/-- The Legendre symbol `legendreSym p a = 1` if there is a solution in `ℤ/pℤ`
+of the equation `x^2 - a*y^2 = 0` with `x ≠ 0`. -/
+theorem eq_one_of_sq_sub_mul_sq_eq_zero' {p : ℕ} [Fact p.Prime] {a : ℤ} (ha : (a : ZMod p) ≠ 0)
+    {x y : ZMod p} (hx : x ≠ 0) (hxy : x ^ 2 - a * y ^ 2 = 0) : legendreSym p a = 1 := by
+  haveI hy : y ≠ 0 := by
+    rintro rfl
+    rw [zero_pow' 2 (by norm_num), MulZeroClass.mul_zero, sub_zero, pow_eq_zero_iff
+        (by norm_num : 0 < 2)] at hxy
+  -- porting note: apparently something was not automatically inferred and now is.
+    exact hx hxy
+  exact eq_one_of_sq_sub_mul_sq_eq_zero ha hy hxy
+#align legendre_sym.eq_one_of_sq_sub_mul_sq_eq_zero' legendreSym.eq_one_of_sq_sub_mul_sq_eq_zero'
+
+/-- If `legendreSym p a = -1`, then the only solution of `x^2 - a*y^2 = 0` in `ℤ/pℤ`
+is the trivial one. -/
+theorem eq_zero_mod_of_eq_neg_one {p : ℕ} [Fact p.Prime] {a : ℤ} (h : legendreSym p a = -1)
+    {x y : ZMod p} (hxy : x ^ 2 - a * y ^ 2 = 0) : x = 0 ∧ y = 0 := by
+  have ha : (a : ZMod p) ≠ 0 := by
+    intro hf
+    rw [(eq_zero_iff p a).mpr hf] at h
+    -- porting note `zero_ne_neg_of_ne` not ported? but no noalign? fixed by `tauto`
+    tauto
```

Comment:
I agree we can probably skip porting this one: it was an internal implementation detail, and not that great as a public facing lemma.

`linarith` might be clearer here than `tauto`.

Reply:
done